### ISSUE TITLE
508-defect-1 ❗Launchblocker [AXE-CORE, COGNITION, SEMANTIC MARKUP]: Form fields and their unique IDs MUST NOT be duplicated on the page

### ISSUE THREAD

**Issue by joshkimux** [department-of-veterans-affairs/va.gov-team#19594]
# [508-defect-1 :exclamation: Launchblocker](https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/platform/accessibility/guidance/defect-severity-rubric.md#508-defect-1)

## Feedback framework

- **❗️ Must** for if the feedback must be applied 
- **⚠️ Should** if the feedback is best practice 
- **✔️ Consider** for suggestions/enhancements 

## Definition of done

1. Review and acknowledge feedback.
1. Fix and/or document decisions made.
1. Accessibility specialist will close ticket after reviewing documented decisions / validating fix.

## Point of Contact

**VFS Point of Contact:** _Josh_

## Axe Violations

**[IDs used in ARIA and labels must be unique - Critical](https://dequeuniversity.com/rules/axe/3.5/duplicate-id-aria?application=coconut)**
Duplicate ARIA IDs are common validation errors that may break the accessibility of labels, e.g., form fields, table header cells. Unique ID's differentiate each element from another and prevent invalid markup, wherein only the first instance gets acted upon by client-side scripting, or where assistive technologies typically only reference the first one accurately.

**[Form fields do not have duplicate labels - Moderate](https://dequeuniversity.com/rules/axe/3.5/form-field-multiple-labels?application=coconut)**
Assigning multiple labels to the same form field can cause problems for some combinations of screen readers and browsers, and the results are inconsistent from one combination to the next. Some combinations will read the first label. Some will read the last label. Others will read both labels.

**[ID attribute values must be unique - Minor](https://dequeuniversity.com/rules/axe/3.5/duplicate-id?application=coconut)**
The ID attribute uniquely identifies elements on a page. It does not make sense to duplicate an ID.

Duplicate ID's can break the accessibility of labels for forms, table header cells, etc., by the second instance being skipped by screen readers, or by client-side scripts. They are common markup validation errors that can eliminate possible sources of accessibility problems, when not breaking accessibility.

Those experienced with client-side scripting know that when you re-use an id, typically the only one that gets acted upon by the scripting is the first instance of the use of that ID. Similarly, assistive technologies may, when referencing an id, only reference the first one accurately.

## Details

It seems like fields were duplicated for all the edit labels and fields within disability compensation and education benefits. This resulted in a number of axe violations that must be resolved before launch.

## Acceptance Criteria

- [ ] All axe violations listed are resolved

## Environment

* Operating System: any
* Browser: all
* Screenreading device: any
* Server destination: staging

## Steps to Recreate

1. Enter `https://staging.va.gov/profile/direct-deposit` in browser
2. Open the edit states for both sections
3. Run axe-coconut
4. Confirm "IDs used..." error is flagged for each duplicated field (3)
5. Confirm "id attributes..." error is flagged for labels and error (4)
6. Confirm "Form fields..." error is flagged for all form fields (6)

## Proposed Solution

**[IDs used in ARIA and labels must be unique - Critical](https://dequeuniversity.com/rules/axe/3.5/duplicate-id-aria?application=coconut)**
Rename any duplicate ARIA ID values. Duplicate ARIA IDs are common validation errors that may break the accessibility of labels, e.g., form fields, table header cells. To fix the problem, change an ID value if it is used more than once to be sure each is unique. Unique ID's differentiate each element from another and prevent invalid markup, wherein only the first instance gets acted upon by client-side scripting, or where assistive technologies typically only reference the first one accurately.

**[Form fields do not have duplicate labels - Moderate](https://dequeuniversity.com/rules/axe/3.5/form-field-multiple-labels?application=coconut)**
Ensure that only one label is assigned to a form field.

**[ID attribute values must be unique - Minor](https://dequeuniversity.com/rules/axe/3.5/duplicate-id?application=coconut)**
Rename any duplicate ID attributes values.

Duplicate IDs are common validation errors that may break the accessibility of labels, e.g., form fields, table header cells.

To fix the problem, change an ID value if it is used more than once to be sure each is unique. Unique ID's differentiate each element from another and prevent invalid markup, wherein only the first instance gets acted upon by client-side scripting, or where assistive technologies typically only reference the first one accurately.

Proper markup eliminates at least one possible source of accessibility problems. WCAG 1.0 used to have a provision which explicitly required the use of valid markup. WCAG 2.0 no longer requires valid markup. We mention it here because valid markup is a quick path toward ensuring accessibility. There are hundreds of ways a person can create invalid markup. Most of these ways are usually inconsequential for accessibility such as un-encoded ampersands, for instance. Others, however, are significant and may cause issues relating to how assistive technology interacts with the page and renders the page to the end user, such as subsequent instances of element attribute ID's being ignored.

Submit markup code to the W3C validator at http://validator.w3.org to the validity of HTML markup in order to quickly identify reused attribute ID values.

## WCAG or Vendor Guidance

* [IDs used in ARIA and labels must be unique - Critical](https://dequeuniversity.com/rules/axe/3.5/duplicate-id-aria?application=coconut)
* [Form fields do not have duplicate labels - Moderate](https://dequeuniversity.com/rules/axe/3.5/form-field-multiple-labels?application=coconut)
* [ID attribute values must be unique - Minor](https://dequeuniversity.com/rules/axe/3.5/duplicate-id?application=coconut)

## Screenshots or Trace Logs

<img width="1468" alt="Screen Shot 2021-02-08 at 11 00 17 AM" src="https://user-images.githubusercontent.com/14154792/107245196-dfa8f880-69fc-11eb-8732-0d689230aa91.png">
<img width="1336" alt="Screen Shot 2021-02-08 at 11 09 31 AM" src="https://user-images.githubusercontent.com/14154792/107246975-ce60eb80-69fe-11eb-9471-85bf14b4b711.png">

**Comment by erikphansen** [comment]
This'll be easy enough to fix. We are reusing the same schema and ui schemas for both bank info edit forms.

We should:
- add a new prop, maybe called `formFieldPrefix`, to the `BankInfoForm` https://github.com/department-of-veterans-affairs/vets-website/blob/234610f5011e301bc05bbd64ef52c21cedbb559f/src/applications/personalization/profile/components/direct-deposit/BankInfoForm.jsx#L52-L59
- make a new function that takes a single arg (the above prop value) and returns a form schema and ui schema, computing each prop value from the passed-in arg.
- call that function within the `BankInfoForm` to get the schema and ui schema that can be passed to the `SchemaForm` https://github.com/department-of-veterans-affairs/vets-website/blob/234610f5011e301bc05bbd64ef52c21cedbb559f/src/applications/personalization/profile/components/direct-deposit/BankInfoForm.jsx#L65-L66

**Comment by erikphansen** [comment]
Actually slightly more complicated. Each form's [submit handler](https://github.com/department-of-veterans-affairs/vets-website/blob/566cbc2a7b3acb6afc35b248aab6cf47515b4571/src/applications/personalization/profile/components/direct-deposit/BankInfoCNPv2.jsx#L88-L93) will need to be updated to correctly parse the form data since each schema property will have a new prefix.
